File: Data Structures & Algorithms/course-schedule-ii/submission-14.ts
class Solution {
    /**
     * @param {number} numCourses
     * @param {number[][]} prerequisites
     * @return {number[]}
     */
    findOrder(numCourses: number, prerequisites: number[][]): number[] {
        const adjList = new Map()
        const visiting = new Set()
        const completed = new Set<number>()

        for (let i = 0; i < numCourses; i++) {
            adjList.set(i, [])
        }

        for (const [course, pre] of prerequisites) {
            adjList.get(course).push(pre)
        }

        const dfs = (course: number) => {
            if (completed.has(course)) return true
            if (visiting.has(course)) return false

            visiting.add(course)

            for (const n of adjList.get(course)) {
                if (!dfs(n)) return false
            }

            completed.add(course)
            visiting.delete(course)

            return true
        }

        for (let i = 0; i < numCourses; i++) {
            // if (completed.has(i)) continue
            if (!dfs(i)) return []
        }

        return [...completed]
    }
}

File: Data Structures & Algorithms/course-schedule/submission-14.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adjList = {i: [] for i in range(numCourses)}
        visiting: set[int] = set()
        completed: set[int] = set()

        # add pre to course
        for c, p in prerequisites:
            adjList[c].append(p)

        
        def dfs(course: int): 
            if course in completed: return True
            if course in visiting: return False

            visiting.add(course)

            for pre in adjList[course]:
                if not dfs(pre): return False

            completed.add(course)
            visiting.remove(course)
            return True

        for i in range(numCourses):
            print(i)
            if not dfs(i): return False

        return True
        